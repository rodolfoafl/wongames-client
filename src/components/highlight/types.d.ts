export type HighlightProps = {
  title: string
  subtitle: string
  coverImage: string
  floatImage?: string
  alignment?: 'left' | 'right'
  children?: React.ReactNode
}
