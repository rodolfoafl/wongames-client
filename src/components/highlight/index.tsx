import { HighlightProps } from '@/components/highlight/types'
import {
  floatImageVariants,
  sectionVariants,
  textContentVariants
} from '@/components/highlight/variants'
import Image from 'next/image'

export default function Highlight({
  title,
  subtitle,
  coverImage,
  floatImage,
  alignment = 'right',
  children
}: HighlightProps) {
  return (
    <section
      className={sectionVariants({ hasFloatImage: !!floatImage, alignment })}
    >
      <Image
        src={coverImage}
        alt={title}
        width={1040}
        height={320}
        title="Cover Image"
        aria-label={title}
        className="absolute inset-0 h-[14.375rem] w-full object-cover object-center md:h-80"
      />
      {!!floatImage && (
        <Image
          src={floatImage}
          alt={title}
          width={320}
          height={320}
          title="Float Image"
          className={floatImageVariants({ alignment })}
        />
      )}
      <div className={textContentVariants({ alignment })}>
        <h2 className="text-lg leading-none font-semibold text-white md:text-[1.75rem]">
          {title}
        </h2>
        <h3 className="text-sm leading-none font-light text-white md:text-lg">
          {subtitle}
        </h3>
        {!!children && <div className="mt-8">{children}</div>}
      </div>
    </section>
  )
}
