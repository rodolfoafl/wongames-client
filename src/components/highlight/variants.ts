import { cn } from '@/utils'
import { tv } from 'tailwind-variants'

export const sectionVariants = tv({
  base: cn(
    'relative grid h-[14.375rem] grid-cols-1',
    'after:absolute after:h-full after:w-full after:bg-[#00000099]',
    'md:h-80'
  ),
  variants: {
    hasFloatImage: {
      true: ''
    },
    alignment: {
      left: '',
      right: ''
    }
  },
  compoundVariants: [
    {
      hasFloatImage: true,
      alignment: 'left',
      className: 'grid-cols-[2fr_1.4fr]'
    },
    {
      hasFloatImage: true,
      alignment: 'right',
      className: 'grid-cols-[1.4fr_2fr]'
    }
  ]
})

export const floatImageVariants = tv({
  base: 'z-10 max-h-[14.375rem] self-end object-contain md:h-80 md:max-h-80',
  variants: {
    alignment: {
      left: 'order-2 justify-self-end',
      right: 'order-1'
    }
  }
})

export const textContentVariants = tv({
  base: 'z-10 flex flex-col gap-y-2 p-4 md:self-end md:p-10',
  variants: {
    alignment: {
      left: 'order-1 items-start text-left',
      right: 'order-2 items-end text-right'
    }
  }
})
