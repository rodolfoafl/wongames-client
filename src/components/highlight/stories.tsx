import { Meta, StoryObj } from '@storybook/react'
import Highlight from '.'
import { HighlightProps } from '@/components/highlight/types'
import LinkButton from '@/components/link-button'

export default {
  title: 'Highlight',
  component: Highlight,
  args: {
    title: 'Red Dead is back',
    subtitle: 'Come see Johnâ€™s new adventures',
    coverImage: '/img/red-dead-img.jpg',
    alignment: 'right',
    children: (
      <LinkButton href="/buy-now" size="large">
        Buy now
      </LinkButton>
    )
  },
  argTypes: {
    alignment: {
      options: ['left', 'right'],
      control: { type: 'radio' }
    }
  },
  parameters: {
    layout: 'fullscreen'
  }
} as Meta

export const Default: StoryObj<HighlightProps> = {
  render: (args) => (
    <div className="mx-auto max-w-[65rem]">
      <Highlight {...args} />
    </div>
  )
}

export const WithFloatImage: StoryObj<HighlightProps> = {
  args: {
    floatImage: '/img/red-dead-float.png',
    alignment: 'right'
  },
  argTypes: {
    alignment: {
      options: ['left', 'right'],
      control: { type: 'radio' }
    }
  },
  render: (args) => (
    <div className="mx-auto max-w-[65rem]">
      <Highlight {...args} />
    </div>
  )
}
