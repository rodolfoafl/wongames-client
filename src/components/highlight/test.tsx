import { render, screen } from '@testing-library/react'

import Highlight from '.'
import LinkButton from '@/components/link-button'
import { HighlightProps } from '@/components/highlight/types'

const props: HighlightProps = {
  title: 'Heading 1',
  subtitle: 'Heading 2',
  coverImage: '/img/red-dead-img.jpg',
  children: <LinkButton href="/link">Buy now</LinkButton>
}

describe('<Highlight />', () => {
  it('should render the cover image, title and subtitle', () => {
    const { container } = render(<Highlight {...props} />)

    const coverImage = screen.getByTitle(/cover image/i)
    expect(coverImage).toBeInTheDocument()
    expect(coverImage).toHaveAttribute('src')

    expect(
      screen.getByRole('heading', { name: props.title })
    ).toBeInTheDocument()
    expect(
      screen.getByRole('heading', { name: props.subtitle })
    ).toBeInTheDocument()
    expect(screen.getByRole('link', { name: /buy now/i })).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render the float image', () => {
    render(<Highlight {...props} floatImage={'/img/red-dead-float.png'} />)

    const coverImage = screen.getByTitle(/float image/i)
    expect(coverImage).toBeInTheDocument()
    expect(coverImage).toHaveAttribute('src')
  })

  it('should render with the text aligned right by default', () => {
    render(<Highlight {...props} floatImage="/img/red-dead-float.png" />)

    const coverImage = screen.getByTitle(/float image/i)
    expect(coverImage).toHaveClass('order-1')

    const textContentParent = screen.getByRole('heading', {
      name: props.title
    }).parentElement
    expect(textContentParent).toHaveClass('order-2')
  })

  it('should render with the text aligned left', () => {
    render(
      <Highlight
        {...props}
        floatImage="/img/red-dead-float.png"
        alignment="left"
      />
    )

    const coverImage = screen.getByTitle(/float image/i)
    expect(coverImage).toHaveClass('order-2')

    const textContentParent = screen.getByRole('heading', {
      name: props.title
    }).parentElement
    expect(textContentParent).toHaveClass('order-1')
  })
})
