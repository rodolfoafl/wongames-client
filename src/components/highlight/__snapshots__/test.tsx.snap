// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Highlight /> should render the cover image, title and subtitle 1`] = `
<section
  class="relative grid h-[14.375rem] grid-cols-1 after:absolute after:h-full after:w-full after:bg-[#00000099] md:h-80"
>
  <img
    alt="Heading 1"
    aria-label="Heading 1"
    class="absolute inset-0 h-[14.375rem] w-full object-cover object-center md:h-80"
    data-nimg="1"
    decoding="async"
    height="320"
    loading="lazy"
    src="/_next/image?url=%2Fimg%2Fred-dead-img.jpg&w=3840&q=75"
    srcset="/_next/image?url=%2Fimg%2Fred-dead-img.jpg&w=1080&q=75 1x, /_next/image?url=%2Fimg%2Fred-dead-img.jpg&w=3840&q=75 2x"
    style="color: transparent;"
    title="Cover Image"
    width="1040"
  />
  <div
    class="z-10 flex flex-col gap-y-2 p-4 md:self-end md:p-10 order-2 items-end text-right"
  >
    <h2
      class="text-lg leading-none font-semibold text-white md:text-[1.75rem]"
    >
      Heading 1
    </h2>
    <h3
      class="text-sm leading-none font-light text-white md:text-lg"
    >
      Heading 2
    </h3>
    <div
      class="mt-8"
    >
      <a
        class="flex items-center justify-center rounded-sm border-none bg-[linear-gradient(180deg,#FF5F5F_0%,#F062C0_100%)] w-fit cursor-pointer p-2 font-semibold text-white hover:bg-[linear-gradient(180deg,#E35565_0%,#D958A6_100%)] h-10 px-4 text-sm"
        href="/link"
      >
        Buy now
      </a>
    </div>
  </div>
</section>
`;
